src/js/smtpserver.c

```diff
@@ -121,41 +452,452 @@ int init_header_class(JSContext *cx, JSObject *global) {
+	};
+
+	JSObject *smtpResponseClass;
+
+	// Create the SmtpPath class
+	smtpResponseClass = JS_InitClass(cx, global, NULL, &smtp_response_class, response_construct, 1, NULL, NULL, NULL, NULL);
+
+	if (!smtpResponseClass) {
 		return -1;
 	}
 
-	if (!JS_DefineProperty(cx, proto, "parts", OBJECT_TO_JSVAL(parts), NULL, NULL, JSPROP_ENUMERATE | JSPROP_READONLY | JSPROP_PERMANENT)) {
+	return 0;
+}
+
+static int connect_to_address(char *ip, char *port)
+{
+	int sockfd, status, portno;
+	struct sockaddr_in serv_addr;
+	struct hostent *server;
+
+	portno = atoi(port);
+	sockfd = socket(AF_INET, SOCK_STREAM, 0);
+
+	if (sockfd < 0) {
+		// throw exc
+		printf("sock failed\n");
 		return -1;
 	}
 
-	// Define other methods
-	if (!JS_DefineFunction(cx, proto, "getValue", header_getValue, 0, 0)) {
+	server = gethostbyname(ip);
+
+	if (!server) {
+		// throw exc
+		printf("server failed\n");
+	}
+
+	bzero((char*) &serv_addr, sizeof(serv_addr));
+
+	serv_addr.sin_family = AF_INET;
+
+	bcopy((char *) server->h_addr, (char *) &serv_addr.sin_addr.s_addr, server->h_length);
+	serv_addr.sin_port = htons(portno);
+
+	if (connect(sockfd, (struct sockaddr *) &serv_addr, sizeof(serv_addr)) < 0) {
+		JS_ReportError(js_context, "Cannot connect to %s:%s!", ip, port);
 		return -1;
 	}
 
-	if (!JS_DefineFunction(cx, proto, "toString", header_toString, 0, 0)) {
+	return sockfd;
+}
+
+static JSBool smtpClient_connect(JSContext *cx, unsigned argc, jsval *vp) {
+	jsval host, port, client, connection, clientStream;
+	char *c_host, *c_port;
+	int sockfd;
+	bfd_t *client_stream;
+
+	client = JS_THIS(cx, vp);
+
+	// Get host
+	if (!JS_GetProperty(cx, JSVAL_TO_OBJECT(client), "host", &host)) {
+		return JS_FALSE;
+	}
+
+	// Get port
+	if (!JS_GetProperty(cx, JSVAL_TO_OBJECT(client), "port", &port)) {
+		return JS_FALSE;
+	}
+
+	c_host = JS_EncodeString(cx, JSVAL_TO_STRING(host));
+	c_port = JS_EncodeString(cx, JSVAL_TO_STRING(port));
+
+	sockfd = connect_to_address(c_host, c_port);
+
+	client_stream = bfd_alloc(sockfd);
+	clientStream = PRIVATE_TO_JSVAL(client_stream);
+
+	if (!JS_SetProperty(cx, JSVAL_TO_OBJECT(client), "clientStream", &clientStream)) {
+		return JS_FALSE;
+	}
+
+	free(c_host);
+	free(c_port);
+
+	return JS_TRUE;
+}
+
+static JSBool smtpClient_readResponse(JSContext *cx, unsigned argc, jsval *vp) {
+	jsval smtpClient, connection, content, response, clientStream;
+	jsval js_code, js_messages, js_disconnect;
+	JSObject *messages_obj, *global;
+
+	int sockfd, code, lines_count;
+	char buf[SMTP_COMMAND_MAX + 1], *p, sep;
+	ssize_t sz;
+	bfd_t *client_stream;
+
+	global = JS_GetGlobalForScopeChain(cx);
+	smtpClient = JS_THIS(cx, vp);
+	messages_obj = JS_NewArrayObject(cx, 0, 0);
+
+	if (!JS_GetProperty(cx, JSVAL_TO_OBJECT(smtpClient), "clientStream", &clientStream)) {
+		return JS_FALSE;
+	}
+
+	client_stream = JSVAL_TO_PRIVATE(clientStream);
+
+	sockfd = client_stream->fd;
+
+	lines_count = 0;
+	do {
+		sz = 0;
+		do {
+			if (read(sockfd, &buf[sz++], 1) < 0)
```

Comment:
Reading from client_stream->fd directly is asking for trouble. The problem is that the bfd_t functions may have already consumed data from the fd and buffered it inside the bfd_t structure. You should use the bfd_t API instead.